ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRewriteToDataSketchesRules.java

```diff
@@ -368,4 +389,216 @@ void rewrite(AggregateCall aggCall) {
       }
     }
   }
+
+  /**
+   * Generic support for rewriting Windowing expression into a different form usually using joins.
+   */
+  private static abstract class WindowingToProjectAggregateJoinProject extends RelOptRule {
+
+    protected final String sketchType;
+
+    public WindowingToProjectAggregateJoinProject(String sketchType) {
+      super(operand(HiveProject.class, any()));
+      this.sketchType = sketchType;
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+
+      final Project project = call.rel(0);
+
+      VbuilderPAP vb = buildProcessor(call);
+      RelNode newProject = vb.processProject(project);
+
+      if (newProject instanceof Project && ((Project) newProject).getChildExps().equals(project.getChildExps())) {
```

Comment:
I think this check could be done more efficiently, avoiding traversing all the expressions.

We need to make sure we only create a new project when there are rewritings in the RexNode expressions:
1) In the `for` loop in `processProject`, check for each expression whether it was modified by ProcessShuttle (`==`).
2) Only create a new Project operator if any of the expressions changed.
3) Check here `newProject == project`.

Reply:
done it a bit differently - but the end result is the same